### DISCUSSION TITLE
Connection pool in tonic?

### DISCUSSION THREAD

**Discussion by YjyJeff** [hyperium/tonic#1134] [Q&A] [answered]
According to the documentation of [Channel](https://docs.rs/tonic/latest/tonic/transport/struct.Channel.html) : It is based on [hyper::Client], the [hyper::Client] has a connection pool to for better performance. Therefore, we do not need the connection pool for tonic. However, after reading the source code of the [Channel], I found that it is based on the [HttpConnector](https://docs.rs/hyper/0.14.22/hyper/client/connect/struct.HttpConnector.html#method.set_keepalive),  which does not support connection pool, as far as I know....

So, for better performance, do we still need a connection pool for tonic? Maybe using the [hyper::Client] as an underling implementation is better?

Thanks in advance, sincerely~

[hyper::Client]: https://docs.rs/hyper/0.14.22/hyper/client/struct.Client.html
[Channel]: https://docs.rs/tonic/latest/tonic/transport/struct.Channel.html

**Comment by LucioFranco** [comment, marked as answer]
Tonic uses the lower-level client from hyper https://github.com/hyperium/tonic/blob/master/tonic/src/transport/service/connection.rs#L5 and does not use its connection pool.

**Reply by YjyJeff** [reply]
Thank you for your reply~ 

According to the [tonic-wrap](https://github.com/hyperium/tonic/blob/master/examples/src/hyper_warp/client.rs#L16) example, we can use the high-level client from hyper: `hyper::Client` to do the same thing. So, what is the advantage of the lower-level client in tonic? And in which situations should I prefer to use `hyper::Client`? Thanks~

**Reply by LucioFranco** [reply]
It all depends on your setup, the tonic client aligns better with regular gRPC clients while the hyper one is more general.

**Reply by YjyJeff** [reply]
Got it, thanks~
